### PR TITLE
Add trajectory interpolation provider(New)

### PR DESCRIPTION
**Pull request by valaxkong**
Follow up issue #400

Add test_controller_interpolation.py

### CODE REVIEW

**Comment by Gamenot** [comment]
@valaxkong Thank you for the contribution. I will tag in @iman512003 and get this merged in.

File: smarts/core/controllers/trajectory_interpolation_controller.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        else:
+            ms = cls.interpolate(ms0, ms1, sim.timestep_sec)
+
+            center_position = ms[X_INDEX : Y_INDEX + 1]
+            center_heading = Heading(ms[THETA_INDEX])
+            pose = Pose.from_center(center_position, center_heading)
+            speed = ms[VEL_INDEX]
+
+        vehicle.set_pose(pose)
+        vehicle.set_speed(speed)
```
Comment (Gamenot) [inline smarts/core/controllers/trajectory_interpolation_controller.py:127]: With the `BoxChassis` this then becomes:
```suggestion
        vehicle.set_control(pose, speed)
```

File: smarts/core/controllers/trajectory_interpolation_controller.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import math
+import numpy as np
+
+from smarts.core.coordinates import Heading, Pose
+from smarts.core.chassis import AckermannChassis
```
Comment (Gamenot) [inline smarts/core/controllers/trajectory_interpolation_controller.py:24]:
```suggestion
from smarts.core.chassis import BoxChassis
```

**Comment by valaxkong** [comment]
Sorry for leaking these bugs into this PR.
Actually, I hit a snag when I tried to run unit tests on my machine.
I tried to run unit test with:
`python -m pytest smarts/core/tests/test_controller_interpolation.py `

And I got this:
`=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.7.9, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /root/SMARTS
plugins: forked-1.3.0, xdist-2.2.1, benchmark-3.2.3, cov-2.11.1, notebook-0.6.1
collected 7 items                                                                                                                                                                                                  

smarts/core/tests/test_controller_interpolation.py .Fatal Python error: Segmentation fault

Current thread 0x00007f47cd3b1740 (most recent call first):
  File "/usr/lib/python3.7/abc.py", line 127 in __new__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/_interceptor.py", line 150 in <module>
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 728 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1035 in _handle_fromlist
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/_server.py", line 28 in <module>
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 728 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1035 in _handle_fromlist
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/__init__.py", line 1992 in server
  File "/root/SMARTS/smarts/zoo/manager.py", line 39 in serve
  File "/usr/lib/python3.7/multiprocessing/process.py", line 99 in run
  File "/usr/lib/python3.7/multiprocessing/process.py", line 297 in _bootstrap
  File "/usr/lib/python3.7/multiprocessing/popen_fork.py", line 74 in _launch
  File "/usr/lib/python3.7/multiprocessing/popen_fork.py", line 20 in __init__
  File "/usr/lib/python3.7/multiprocessing/context.py", line 277 in _Popen
  File "/usr/lib/python3.7/multiprocessing/context.py", line 223 in _Popen
  File "/usr/lib/python3.7/multiprocessing/process.py", line 112 in start
  File "/root/SMARTS/smarts/core/remote_agent_buffer.py", line 189 in spawn_local_zoo_manager
  File "/root/SMARTS/smarts/core/remote_agent_buffer.py", line 77 in __init__
  File "/root/SMARTS/smarts/core/agent_manager.py", line 46 in __init__
  File "/root/SMARTS/smarts/core/smarts.py", line 152 in __init__
  File "/root/SMARTS/smarts/core/tests/test_controller_interpolation.py", line 196 in smarts
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 925 in call_fixture_func
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 1126 in pytest_fixture_setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 1072 in execute
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 687 in _compute_fixture_value
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 601 in _get_active_fixturedef
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 581 in getfixturevalue
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 568 in _fillfixtures
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/python.py", line 1647 in setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 449 in prepare
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 150 in pytest_runtest_setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 311 in from_call
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 255 in call_runtest_hook
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 120 in runtestprotocol
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 323 in _main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/config/__init__.py", line 163 in main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/lib/python3.7/runpy.py", line 85 in _run_code
  File "/usr/lib/python3.7/runpy.py", line 193 in _run_module_as_main

^Cargv[0]=
:display:x11display(fatal): X fatal error on display 0x5625ba0d4f90
`
Do you know what happened on this?

@Gamenot

**Comment by valaxkong** [comment]
Sorry for leaking these bugs into this PR.
Actually, I hit a snag when I tried to run unit tests on my machine.
I tried to run unit test with:
`python -m pytest smarts/core/tests/test_controller_interpolation.py `

And I got this:
`=============================================================================================== test session starts ================================================================================================
platform linux -- Python 3.7.9, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /root/SMARTS
plugins: forked-1.3.0, xdist-2.2.1, benchmark-3.2.3, cov-2.11.1, notebook-0.6.1
collected 7 items                                                                                                                                                                                                  

smarts/core/tests/test_controller_interpolation.py .Fatal Python error: Segmentation fault

Current thread 0x00007f47cd3b1740 (most recent call first):
  File "/usr/lib/python3.7/abc.py", line 127 in __new__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/_interceptor.py", line 150 in <module>
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 728 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1035 in _handle_fromlist
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/_server.py", line 28 in <module>
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 728 in exec_module
  File "<frozen importlib._bootstrap>", line 677 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 967 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 983 in _find_and_load
  File "<frozen importlib._bootstrap>", line 219 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1035 in _handle_fromlist
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/grpc/__init__.py", line 1992 in server
  File "/root/SMARTS/smarts/zoo/manager.py", line 39 in serve
  File "/usr/lib/python3.7/multiprocessing/process.py", line 99 in run
  File "/usr/lib/python3.7/multiprocessing/process.py", line 297 in _bootstrap
  File "/usr/lib/python3.7/multiprocessing/popen_fork.py", line 74 in _launch
  File "/usr/lib/python3.7/multiprocessing/popen_fork.py", line 20 in __init__
  File "/usr/lib/python3.7/multiprocessing/context.py", line 277 in _Popen
  File "/usr/lib/python3.7/multiprocessing/context.py", line 223 in _Popen
  File "/usr/lib/python3.7/multiprocessing/process.py", line 112 in start
  File "/root/SMARTS/smarts/core/remote_agent_buffer.py", line 189 in spawn_local_zoo_manager
  File "/root/SMARTS/smarts/core/remote_agent_buffer.py", line 77 in __init__
  File "/root/SMARTS/smarts/core/agent_manager.py", line 46 in __init__
  File "/root/SMARTS/smarts/core/smarts.py", line 152 in __init__
  File "/root/SMARTS/smarts/core/tests/test_controller_interpolation.py", line 196 in smarts
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 925 in call_fixture_func
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 1126 in pytest_fixture_setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 1072 in execute
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 687 in _compute_fixture_value
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 601 in _get_active_fixturedef
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 581 in getfixturevalue
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/fixtures.py", line 568 in _fillfixtures
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/python.py", line 1647 in setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 449 in prepare
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 150 in pytest_runtest_setup
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 255 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 311 in from_call
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 255 in call_runtest_hook
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 215 in call_and_report
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 120 in runtestprotocol
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/runner.py", line 109 in pytest_runtest_protocol
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 348 in pytest_runtestloop
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 323 in _main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 269 in wrap_session
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/main.py", line 316 in pytest_cmdline_main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/config/__init__.py", line 163 in main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/_pytest/config/__init__.py", line 185 in console_main
  File "/root/SMARTS/.venv/lib/python3.7/site-packages/pytest/__main__.py", line 5 in <module>
  File "/usr/lib/python3.7/runpy.py", line 85 in _run_code
  File "/usr/lib/python3.7/runpy.py", line 193 in _run_module_as_main

^Cargv[0]=
:display:x11display(fatal): X fatal error on display 0x5625ba0d4f90
`
Do you know what happened on this?

@Gamenot

**Comment by Gamenot** [comment]
@valaxkong  Very sorry for the extremely late reply. The last few weeks have been rough. I intend to have this merged in by Friday. 

There is no problem with the bugs. They were expected due to the difference in version.

The problem from the tests looks like it may be an older issue we had with gRPC interacting badly with `pytest`.

File: smarts/core/tests/test_controller_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        delta = np.array(
+            [
+                [0.0],
+                [curr_position[0]],
+                [curr_position[1]],
+                [curr_heading],
+                [curr_speed],
+            ]
+        )
+        return delta + self._local_traj
```
[the remark this answers is not in the document]
Comment (valaxkong) [reply smarts/core/tests/test_controller_interpolation.py:153]: I think the codes here agree with your comment. The problem is 'delta' here is confusing. Actually, it is a new local origin state. I correct this at new commit, please take a look.

[the remark this answers is not in the document]
Comment (valaxkong) [reply smarts/core/tests/test_controller_interpolation.py:153]: @Gamenot I think the codes here agree with your comment. The problem is 'delta' here is confusing. Actually, it is a new local origin state. I correct this at new commit, please take a look.

**Comment by valaxkong** [comment]
@Gamenot I use force-push again on this PR according with your comment. Do you prefer a series of submissions or just one submission for one issue? Will you squash commits before merging them into master branch?

**Comment by valaxkong** [comment]
@Gamenot 
After reading some relative code further, I rewrite this PR.
I implement a new provider, trajectory interpolation provider, to achieve my previous purpose. I think it will be more in line with the design spirit of SMARTS.
Please take a look.

**Review by liu-allan — APPROVED** [review]

**Comment by Gamenot** [comment]
Everything here looks good now.

File: smarts/core/agent_interface.py [hunk 4]
```diff
@@ -127,6 +127,8 @@ class AgentType(IntEnum):
     """Controls multiple vehicles"""
     MPCTracker = 10
     """Agent performs trajectory tracking using model predictive control."""
+    TrajectoryInterpolator = 11
+    """Agent performs linear trajectory interpolation."""
     Imitation = 11
```
Comment (sah-huawei) [inline smarts/core/agent_interface.py:132]: Merge error:  Note that 11 is used by two things now.

File: smarts/core/controllers/__init__.py [hunk 5]
```diff
@@ -48,6 +48,7 @@ class ActionSpaceType(Enum):
     Trajectory = 5
     MultiTargetPose = 6  # for boid control
     MPC = 7
+    TrajectoryWithTime = 8  # for pure interpolation provider
     Imitation = 8
```
Comment (sah-huawei) [inline smarts/core/controllers/__init__.py:52]: And here, 8 is used by two things.

**Review by sah-huawei — COMMENTED** [review]
There were two merge errors when this PR was merged into the develop branch.  They should be corrected immediately.

File: smarts/core/agent_interface.py (hunk 4 above)
Comment (RutvikGupta) [reply smarts/core/agent_interface.py:132]: Oh thanks for letting me know. Will resolve it.

Comment (RutvikGupta) [reply smarts/core/agent_interface.py:132]: Opened PR #978  to fix it.

**Pull request merged**